bcipy/language/README.md

```diff
@@ -16,4 +16,33 @@ The pretrained GPT2 language model is saved in [this folder on Google Drive](htt
\ No newline at end of file
+The UniformLanguageModel provides equal probabilities for all symbols in the symbol set. This model is useful for evaluating other aspects of the system, such as EEG signal quality, without strong influence from a language model.
+
+## GPT2 Model
+
+The GPT2LanguageModel utilizes Huggingface's pretrained GPT2 language model and the beam search algorithm to generate probabilities for all symbols in the symbol set. This is the model we are currently using to incorporate language model predictions into BciPy experiments.
+
+For the GPT2 model, byte-pair encoding (BPE) is used for tokenization. The main idea of BPE is to create a fixed-size vocabulary that contains common English subword units. Then a less common word would be broken down into several subword units in the vocabulary. For example, the tokenization of character sequence `peanut_butter_and_jel` would be:
+> *['pe', 'anut', '_butter', '_and', '_j', 'el']*
+
+Therefore, in order to generate a predictive distribution on the next character, we need to examine all the possibilities that could complete the final subword units in the input sequences. We simply take the input and remove the last (partially-typed) subword unit and make the model predict it over the entire vocabulary. Then we select the subword units with prefixes matching the partially-typed last subword from the input (in the example case above, the selected subword units should start with *'el'*). Then we renormalize the probability distribution over the selected subword units and marginalize over the first character after the matched prefix. The resulting character-level distribution would be a prediction of the next character given current input.
+
+However, predicting only the final subword units may limit the space of possible continuations of the input. To remedy that, we need to look beyond one subword unit. In such occasion, we would utilize beam search to generate multiple subword units. Beam search is a heuristic search algorithm that expands the most promising hypotheses in a selected set. In our case, the most promising hypotheses would be the top ranked subword units on the current beam. For each of these hypotheses, we would make the language model further predict the next subword unit over the entire vocabulary. Then the top ranked hypotheses returned would again be retained for further expansion. The renormalization and marginalization would be performed at the end of the beam search for all candidates on the beam. Note that the beam size and the beam search depth are hyperparameters that can be adjusted.
+
+Furthermore, we use a unigram language model trained on the ALS Phraseset to interpolate with GPT2 for smoothing purpose, since the distributions given by GPT2 can be very "sharp", where most of the probability mass falls on the top few letters. Therefore, interpolating with a unigram model could give more probability to the lower ranked letters.
```

Comment:
We should comment more on this `ALS Phraseset`. Where can it be found? Or if it's from our group, a little more background on its collection!

Reply:
I got the ALS Phraseset from Betts. She should be able to provide more details :)